File: JULY-2025/20-07-2025/p1948.swift
import Foundation

class Node {
    var name: String
    var children: [String: Node] = [:]
    var parent: Node? = nil
    var shouldDelete = false
    
    init(_ name: String = "") {
        self.name = name
    }
}

class Solution {
    func deleteDuplicateFolder(_ paths: [[String]]) -> [[String]] {
        let root = Node()
        
        for path in paths {
            var curr = root
            for name in path {
                if curr.children[name] == nil {
                    let newNode = Node(name)
                    newNode.parent = curr
                    curr.children[name] = newNode
                }
                curr = curr.children[name]!
            }
        }

        var serialToNodes: [String: [Node]] = [:]

        @discardableResult
        func serialize(_ node: Node) -> String {
            if node.children.isEmpty { return "" }
            var result = [String]()
            for key in node.children.keys.sorted() {
                let child = node.children[key]!
                let serialized = serialize(child)
                result.append("(\(key)\(serialized))")
            }
            let final = result.joined()
            serialToNodes[final, default: []].append(node)
            return final
        }

        serialize(root)

        for nodes in serialToNodes.values where nodes.count > 1 {
            for node in nodes {
                markForDeletion(node)
            }
        }

        func markForDeletion(_ node: Node) {
            node.shouldDelete = true
            for child in node.children.values {
                markForDeletion(child)
            }
        }

        var result = [[String]]()

        func collect(_ node: Node, _ path: [String]) {
            if node !== root && node.shouldDelete { return }
            if node !== root {
                let newPath = path + [node.name]
                result.append(newPath)
                for child in node.children.values {
                    collect(child, newPath)
                }
            } else {
                for child in node.children.values {
                    collect(child, [])
                }
            }
        }

        collect(root, [])
        return result
    }
}

File: JULY-2025/20-07-2025/p1948.py
from collections import defaultdict

class TrieNode:
    def __init__(self):
        self.children = dict()
        self.name = ""
        self.serial = ""
        self.is_deleted = False

class Solution:
    def deleteDuplicateFolder(self, paths):
        root = TrieNode()
        
        # Build trie
        for path in paths:
            node = root
            for folder in path:
                if folder not in node.children:
                    child = TrieNode()
                    child.name = folder
                    node.children[folder] = child
                node = node.children[folder]

        # Map serialized subtree → list of nodes
        serial_map = defaultdict(list)

        def serialize(node):
            if not node.children:
                node.serial = ""
                return ""
            items = []
            for child in sorted(node.children.values(), key=lambda x: x.name):
                items.append(f"{child.name}({serialize(child)})")
            node.serial = "".join(items)
            serial_map[node.serial].append(node)
            return node.serial

        serialize(root)

        # Mark duplicates
        for nodes in serial_map.values():
            if len(nodes) > 1:
                for node in nodes:
                    node.is_deleted = True

        # Collect surviving paths
        ans = []

        def collect(node, path):
            for child in node.children.values():
                if not child.is_deleted:
                    new_path = path + [child.name]
                    ans.append(new_path)
                    collect(child, new_path)

        collect(root, [])
        return ans